add lint id enum (#426)

fabricio7p opened the pull request:
Began refactoring `src/checks.rs`

- add child module check_variants
- add Lint variants
- refactor test setup to support the variants

Note: this is just a checkpoint to see if i'm on the right track. Code does not compile yet!

mgrachev commented:
@fabricio7p 👋 Thank you for your contribution 👍

Let us know when you've finished so we can review your PR 😉

fabricio7p commented:
Checkpoint.

- add CheckVariants struct that hold the variants vec. (Vec<Lint>)
- refactor `comment.rs` to support output the new CheckVariants struct
- begin refactoring each `/src/fixes` (only first 3 done)

fabricio7p commented:
Refactor done.
Please review @mgrachev 

- Refactored every occurence of string comparisons to support the new LintKind enum.
- Refactored all tests to support enum variants
- Moved Lint and LintKind to new `lints.rs` file
- Add testing to Lint and LintKind

mgrachev commented:
I'm sorry, I'm quite busy now, so I'll try to review your PR by the end of the week 🙂

mgrachev commented on src/lib.rs:
```diff
@@ -9,9 +9,11 @@ pub mod cli;
 mod common;
 mod fixes;
 mod fs_utils;
+mod lints;
```
Let it be `lint_kind`.

fabricio7p replied:
Done

mgrachev commented on src/common/comment.rs:
```diff
@@ -1,11 +1,13 @@
+pub struct Comment {
     disable: bool,
-    pub checks: Vec<&'a str>,
+    pub checks: Lint,
```
Can you use just `Vec<LintKind>`?

fabricio7p replied:
Done

mgrachev reviewed (commented):
@fabricio7p Thank you for your contribution 🚀

I'm sorry for the late reply.
I have left some comments.
Please take a look at them 👀

fabricio7p commented on src/checks/duplicated_key.rs:
```diff
@@ -1,9 +1,10 @@
 
 pub(crate) struct DuplicatedKeyChecker<'a> {
-    name: &'a str,
+    name: LintKind,
```
Done

fabricio7p commented on src/common/comment.rs:
```diff
@@ -85,15 +90,15 @@ mod tests {
             Some(Comment {
                 disable: false,
-                checks: vec![],
+                checks: Lint::new(),
```
Done

fabricio7p commented on src/common/comment.rs:
```diff
@@ -123,15 +128,15 @@ mod tests {
             Some(Comment {
                 disable: true,
-                checks: vec!["UnorderedKey"],
+                checks: Lint::from(vec!["UnorderedKey"]),
```
Done

fabricio7p commented on src/lints.rs:
```diff
@@ -0,0 +1,117 @@
+use std::{fmt, str::FromStr};
+
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub struct Lint {
```
Done

fabricio7p commented on src/fixes.rs:
```diff
@@ -133,7 +137,7 @@ mod tests {
         let mut warnings = vec![Warning::new(
             lines[1].clone(),
-            "Unfixable",
+            LintKind::Unfixable,
```
Done

fabricio7p commented on src/lints.rs:
```diff
@@ -0,0 +1,117 @@
+    SubstitutionKey,
+    TrailingWhitespace,
+    UnorderedKey,
+    Unfixable,
```
Done

fabricio7p commented:
Alright, everything refactored.
Rebased from master as well.

mgrachev commented on src/common/comment.rs:
```diff
@@ -52,6 +63,13 @@ impl Comment<'_> {
 mod tests {
     use super::*;
 
+    fn as_lintkind_vec(vec: Vec<&str>) -> Vec<LintKind> {
```
I don't think that we need a special function in the tests module to convert vectors.
Just use `vec![LintKind::UnorderedKey]`.

mgrachev commented on src/common/line_entry.rs:
```diff
@@ -301,6 +301,16 @@ mod tests {
+        use crate::lint_kind::LintKind;
+        use std::str::FromStr;
+
+        fn as_lintkind_vec(vec: Vec<&str>) -> Vec<LintKind> {
```
I don't think that we need a special function in the tests module to convert vectors.
Just use `vec![LintKind::UnorderedKey]`.

mgrachev commented on src/lib.rs:
```diff
@@ -62,6 +71,11 @@ pub fn fix(args: &clap::ArgMatches, current_dir: &Path) -> Result<(), Box<dyn Er
+        .into_iter()
+        .map(LintKind::from_str)
+        .collect::<Result<Vec<LintKind>, _>>()
+        .unwrap();
```
https://github.com/dotenv-linter/dotenv-linter/pull/427/files#r660372735

mgrachev commented on src/lib.rs:
```diff
@@ -28,6 +31,12 @@ pub fn check(args: &clap::ArgMatches, current_dir: &Path) -> Result<usize, Box<d
+        .into_iter()
+        .map(LintKind::from_str)
+        .collect::<Result<Vec<LintKind>, _>>()
+        .unwrap();
```
https://github.com/dotenv-linter/dotenv-linter/pull/427/files#r660372735

fabricio7p commented:
All good.
Tests passing :+1:

mgrachev commented:
@fabricio7p Thank you!

Please squash your commits to one and fix warnings
https://github.com/dotenv-linter/dotenv-linter/pull/427/checks?check_run_id=2950521632

DDtKey reviewed (approved):
Thank you for very useful PR 🙂

mgrachev commented:
@fabricio7p Thank you for your contribution 👍
